libdeno/modules.cc

```diff
@@ -0,0 +1,205 @@
+
+  v8::EscapableHandleScope handle_scope(isolate);
+
+  auto builtin_modules = d->GetBuiltinModules();
+
+  deno_mod referrer_id = referrer->GetIdentityHash();
+  auto* referrer_info = d->GetModuleInfo(referrer_id);
+  CHECK_NE(referrer_info, nullptr);
+
+  for (int i = 0; i < referrer->GetModuleRequestsLength(); i++) {
+    Local<String> req = referrer->GetModuleRequest(i);
+
+    if (req->Equals(context, specifier).ToChecked()) {
+      v8::String::Utf8Value req_utf8(isolate, req);
+      std::string req_str(*req_utf8);
+
+      deno_mod id = 0;
+      {
+        bool has_builtin = builtin_modules->Has(context, specifier).ToChecked();
+        if (has_builtin) {
+          auto it = d->mods_by_name_.find(req_str.c_str());
+          if (it != d->mods_by_name_.end()) {
+            id = it->second;
+          } else {
+            std::string src = BuiltinModuleSrc(context, specifier);
+            id = d->RegisterModule(req_str.c_str(), src.c_str());
+          }
+        } else {
+          id = d->resolve_cb_(d->user_data_, req_str.c_str(), referrer_id);
+        }
+      }
+
+      auto* info = d->GetModuleInfo(id);
+      if (info == nullptr) {
+        char buf[64 * 1024];
+        snprintf(buf, sizeof(buf), "Cannot resolve module \"%s\" from \"%s\"",
+                 req_str.c_str(), referrer_info->name.c_str());
+        isolate->ThrowException(deno::v8_str(buf, true));
+        break;
+      } else {
+        Local<Module> child_mod = info->handle.Get(isolate);
+        return handle_scope.Escape(child_mod);
+      }
+    }
+  }
+
+  return v8::MaybeLocal<v8::Module>();  // Error
+}
+
+extern "C" {
+
+deno_mod deno_mod_new(Deno* d_, const char* name_cstr,
+                      const char* source_cstr) {
+  auto* d = unwrap(d_);
+  return d->RegisterModule(name_cstr, source_cstr);
+}
+
+const char* deno_mod_name(Deno* d_, deno_mod id) {
+  auto* d = unwrap(d_);
+  auto* info = d->GetModuleInfo(id);
+  return info->name.c_str();
+}
+
+size_t deno_mod_imports_len(Deno* d_, deno_mod id) {
+  auto* d = unwrap(d_);
+  auto* info = d->GetModuleInfo(id);
+  return info->import_specifiers.size();
+}
+
+const char* deno_mod_imports_get(Deno* d_, deno_mod id, size_t index) {
+  auto* d = unwrap(d_);
+  auto* info = d->GetModuleInfo(id);
+  if (info == nullptr || index >= info->import_specifiers.size()) {
+    return nullptr;
+  } else {
+    return info->import_specifiers[index].c_str();
+  }
+}
+
+void deno_mod_instanciate(Deno* d_, void* user_data, deno_mod id,
+                          deno_resolve_cb cb) {
+  auto* d = unwrap(d_);
+  deno::UserDataScope user_data_scope(d, user_data);
+
+  auto* isolate = d->isolate_;
+  v8::Isolate::Scope isolate_scope(isolate);
+  v8::Locker locker(isolate);
+  v8::HandleScope handle_scope(isolate);
+  auto context = d->context_.Get(d->isolate_);
+  v8::Context::Scope context_scope(context);
+
+  v8::TryCatch try_catch(isolate);
+  {
+    // Might need UserDataScope thing.
```

Comment:
Are you calling into rust from here?

Reply:
Poorly worded comment - what I meant is that we might need another Scope class for setting the resolve_cb_ member. I will remove it.